docs/collaborative-pinsets-rfc.md

```diff
@@ -0,0 +1,179 @@
+
+Other components of cluster are independent from these two tasks, and provide functionality that will be useful in scenarios where the peer-set and pin-set maintenance works in a different manner:
+
+* A state component provides pinset representation and serialization
+* An ipfs connector component provides facilities for controlling the ipfs daemon (and the proxy)
+* A pintracker component provides functionality to ensure the shared state is followed by ipfs
+* A monitoring component provides metric collection.
+
+Thus, just replacing the consensus layer (with some caveats ellaborated below) is a relatively simple approach to support collaborative pinsets.
+
+## A new consensus layer for collaborative pinsets
+
+### State of the art of the consensus component
+
+The consensus component is defined by the following interface:
+
+```go
+type Consensus interface {
+    Component
+    Ready() <-chan struct{}
+    LogPin(c api.Pin) error
+    LogUnpin(c api.Pin) error
+    AddPeer(p peer.ID) error
+    RmPeer(p peer.ID) error
+    State() (state.State, error)
+    Leader() (peer.ID, error)
+    WaitForSync() error
+    Clean() error
+    Peers() ([]peer.ID, error)
+}
+```
+
+This is essentially a wrapper of the go-libp2p-consensus which adds functionality and utility methods which until now applied to raft (provided by go-libp2p-raft).
+
+The purpose of Raft is to maintain a shared state by providing a distributed append-only log. In a Raft cluster, the log is maintained by an elected Leader, compacted and snapshotted convieniently. IPFS Cluster has spent significant efforts in detaching the state representation from raft, and allowing transformations (upgrades) to run indepedently, based solely on the "state component".
+
+### A new consensus component using a IPFS-backed CRDTs
+
+In order to be consistent with how components are meant to interact with each-other, it makes sense to implement the shared state maintenance in a new `go-libp2p-consensus` implementation. The main pain points in a collaborative cluster will be:
+
+* Security: all peers should not be able to control or influence the behaviour of all other peers (including updating the state). For this, we will introduce the notion of a *trusted peerset* along with RPC endpoint authorization.
+* Scalability: this consensus layer should support very large number of peers, efficient broadcast of state updates. For this we will introduce a CRDT-based "consensus" layer (it's not consensus per-se) and rely on improved Pubsub mechanisms as provided by libp2p.
+
+### Security: Authentication and authorization for collaborative pinning
+
+In a collaborative pinset scenario, we probably want to have a limited set of peers which are able to modify the shared state and freely connect to API endpoints from any other peers. We call this **trusted peerset**. This implies that we need to find ways to:
```

Comment:
@hsanjuan I just want to confirm my understanding,
A trusted peerset is defined for a given pinset, right? So, A cluster can have multiple trusted peersets based on pinsets. 
One to one mapping between trusted peerset and pinset. Also a cluster can have multiple pinsets as well.

Reply:
Or do you define a trusted peerset first and pinset is defined by pins held by that peerset. i.e, pinset is a function of peerset.

Second one seems more appropriate to me.

Reply:
There is only one pinset per cluster (the shared state).